File: homework2/parrv_jain/Ques - 11 .cpp
// Check if a doubly linked list is a palindrome.
// Method: two-pointer compare from head and tail.
// Time: O(n). Space: O(1).

#include <bits/stdc++.h>
using namespace std;

struct DNode {
    long long val;
    DNode* prev;
    DNode* next;
    DNode(long long v) : val(v), prev(nullptr), next(nullptr) {}
};

DNode* buildDList(const vector<long long>& vals) {
    if (vals.empty()) return nullptr;
    DNode* head = new DNode(vals[0]);
    DNode* cur = head;
    for (size_t i = 1; i < vals.size(); ++i) {
        DNode* node = new DNode(vals[i]);
        cur->next = node;
        node->prev = cur;
        cur = node;
    }
    return head;
}

bool isPalindrome(DNode* head) {
    if (!head) return true;
    // find tail
    DNode* tail = head;
    while (tail->next) tail = tail->next;
    // compare
    while (head != tail && head->prev != tail) {
        if (head->val != tail->val) return false;
        head = head->next;
        tail = tail->prev;
    }
    return true;
}

void freeDList(DNode* head) {
    while (head) {
        DNode* nx = head->next;
        delete head;
        head = nx;
    }
}

int main() {
    ifstream file("input11.txt");
    if (!file) {
        cout << "Error opening file!\n";
        return 1;
    }

    int T;
    if (!(file >> T)) {
        cout << "Invalid input: first token should be the number of test cases.\n";
        return 1;
    }

    for (int tc = 1; tc <= T; ++tc) {
        int n;
        file >> n;
        vector<long long> vals;
        for (int i = 0; i < n; ++i) {
            long long x; file >> x;
            vals.push_back(x);
        }

        DNode* head = buildDList(vals);
        bool ans = isPalindrome(head);

        cout << "Test Case " << tc << ": " << (ans ? "YES" : "NO") << "\n";

        freeDList(head);
    }
    return 0;
}

File: homework2/hansika/ques11.java
/*
Technique Used: Doubly linked list forward-backward two-pointer
Time Complexity: O(n)
Space Complexity: O(1)

Time spent: 30 minutes
*/

class LLNode 
{
    int val;
    LLNode next;
    LLNode prev;
    LLNode(int v) 
    { 
        this.val = v; 
    }
}

public class ques11 
{

    // Check if palindrome
    public static boolean isPalindrome(LLNode head) 
    {
        if (head == null || head.next == null) return true;
        LLNode left = head;
        LLNode right = head;
        // Move right to the tail
        while (right.next != null) 
        {
            right = right.next;
        }
        while (left != right && right.next != left) 
        {
            if (left.val != right.val) return false;
            left = left.next;
            right = right.prev;
        }
        return true;
    }

    // Print helper 
    public static void printList(LLNode head) 
    {
        LLNode t = head;
        while (t != null) 
        {
            System.out.print(t.val);
            if (t.next != null) System.out.print(" <-> ");
            t = t.next;
        }
        System.out.println();
    }

    public static void main(String[] args) 
    {

        //Eg1 (true)
        LLNode a = new LLNode(9);
        a.next = new LLNode(2);     a.next.prev = a;
        a.next.next = new LLNode(4); a.next.next.prev = a.next;
        a.next.next.next = new LLNode(2); a.next.next.next.prev = a.next.next;
        a.next.next.next.next = new LLNode(9); a.next.next.next.next.prev = a.next.next.next;
        System.out.println("Input 1:");
        printList(a);
        System.out.println("Output: " + isPalindrome(a));
        // Eg 2 (False)
        LLNode b = new LLNode(9);
        b.next = new LLNode(12);     b.next.prev = b;
        b.next.next = new LLNode(4); b.next.next.prev = b.next;
        b.next.next.next = new LLNode(2); b.next.next.next.prev = b.next.next;
        b.next.next.next.next = new LLNode(9); b.next.next.next.next.prev = b.next.next.next;
        System.out.println("\nInput 2:");
        printList(b);
        System.out.println("Output: " + isPalindrome(b));
    }
}
